=== modified file 'src/platformsupport/fontdatabases/fontconfig/qfontconfigdatabase.cpp'
--- a/src/platformsupport/fontdatabases/fontconfig/qfontconfigdatabase.cpp	2013-05-03 09:40:08 +0000
+++ b/src/platformsupport/fontdatabases/fontconfig/qfontconfigdatabase.cpp	2015-08-04 11:02:35 +0000
@@ -813,6 +813,11 @@
     return QFont(resolved);
 }
 
+bool QFontconfigDatabase::fontsAlwaysScalable() const
+{
+    return true;
+}
+
 void QFontconfigDatabase::setupFontEngine(QFontEngineFT *engine, const QFontDef &fontDef) const
 {
     bool antialias = !(fontDef.styleStrategy & QFont::NoAntialias);
=== modified file 'src/platformsupport/fontdatabases/fontconfig/qfontconfigdatabase_p.h'
--- a/src/platformsupport/fontdatabases/fontconfig/qfontconfigdatabase_p.h	2013-05-03 09:40:08 +0000
+++ b/src/platformsupport/fontdatabases/fontconfig/qfontconfigdatabase_p.h	2015-08-04 11:03:32 +0000
@@ -57,6 +57,7 @@
     QStringList addApplicationFont(const QByteArray &fontData, const QString &fileName) Q_DECL_OVERRIDE;
     QString resolveFontFamilyAlias(const QString &family) const Q_DECL_OVERRIDE;
     QFont defaultFont() const Q_DECL_OVERRIDE;
+    bool fontsAlwaysScalable() const Q_DECL_OVERRIDE;
 
 private:
     void setupFontEngine(QFontEngineFT *engine, const QFontDef &fontDef) const;
