Description: load testability from environment variable
Author: Thomi Richards <thomi.richards@canonical.com>
Forwarded: https://bugreports.qt-project.org/browse/QTBUG-32974

diff -urN qtbase-opensource-src-5.0.2+dfsg1.old/src/widgets/kernel/qapplication.cpp qtbase-opensource-src-5.0.2+dfsg1/src/widgets/kernel/qapplication.cpp
--- qtbase-opensource-src-5.0.2+dfsg1.old/src/gui/kernel/qguiapplication.cpp	2013-04-09 03:11:06.000000000 +0300
+++ qtbase-opensource-src-5.0.2+dfsg1/src/gui/kernel/qguiapplication.cpp	2013-08-08 10:07:56.982957641 +0300
@@ -1253,6 +1253,9 @@
         argv[j] = 0;
         argc = j;
     }
+
+    if (qgetenv("QT_LOAD_TESTABILITY").toInt() > 0)
+        loadTestability = true;
 
     // Load environment exported generic plugins
     foreach (const QByteArray &plugin, qgetenv("QT_QPA_GENERIC_PLUGINS").split(','))
